wren-ui/src/apollo/server/factories/onboardingFactory.ts

```diff
@@ -297,51 +348,55 @@ export class DuckDBStrategy implements IDataSourceStrategy {
     this.ctx = ctx;
   }
 
-  public async saveDataSource(properties: any) {
-    const { displayName, extensions, configurations } = properties;
-    const initSql = this.concatInitSql(properties.initSql, extensions);
-    const connector = new DuckDBConnector({
-      wrenEngineAdaptor: this.ctx.wrenEngineAdaptor,
-    });
+  public async createDataSource(properties: DuckDBDataSourceProperties) {
+    const { displayName, initSql, extensions, configurations } = properties;
+    const initSqlWithExtensions = this.concatInitSql(initSql, extensions);
 
-    // prepare duckdb environment in wren-engine
-    const prepareOption = {
-      sessionProps: configurations,
-      initSql,
-    } as DuckDBPrepareOptions;
-    await connector.prepare(prepareOption);
-
-    // update wren-engine config
-    const config = {
-      'wren.datasource.type': 'duckdb',
-    };
-    await this.ctx.wrenEngineAdaptor.patchConfig(config);
+    await this.testConnection({
+      initSql: initSqlWithExtensions,
+      configurations,
+    });
 
-    // check DataSource is valid and can connect to it
-    const connected = await connector.connect();
-    if (!connected) {
-      throw new Error('Can not connect to data source');
-    }
-    // check can list dataset table
-    try {
-      await connector.listTables({ format: false });
-    } catch (_e) {
-      throw new Error('Can not list tables in dataset');
-    }
+    await this.patchConfigToWrenEngine();
 
     // save DataSource to database
     const project = await this.ctx.projectRepository.createOne({
       displayName,
       schema: 'public',
       catalog: 'wrenai',
       type: DataSourceName.DUCKDB,
-      initSql,
+      initSql: trim(initSql),
```

Comment:
Just curious, why trim it ?

Reply:
Because the value may contain '/n SELECT xxx from xxx \n\n' etc 🤔️

Reply:
ok noted.